### PR TITLE
Making note of the automatic retry in the docs.

### LINKED ISSUE #831: [2.0 docs] Hook errors advice needs to be updated
Since we now retry hooks on error automatically [in Juju 2.0], this has implications for the advice we give charm developers on [hook errors](https://jujucharms.com/docs/devel/authors-hook-errors).

While nothing really changes in terms of the importance of avoiding hook errors due to charm defects (badly-written charms won't heal themselves!), I think it does have potential impact on how hooks should deal with unexpected externalities such as a temporary network connectivity issue during package installs, etc.

The current docs advise the charm author to "wait a moment and retry a couple of times" in the event of a transient error. I think this advice now adds needless complexity -- it would be better for the hook to fail and retry in such cases where basic assumptions on the state of the system become invalid.

Thoughts? Where do we draw the line on this?

(Thanks @jrwren for bringing this to my attention!)

### PR DESCRIPTION
**Pull request by mbruzek**
Fixes #831 

These are the best words I can find for this awkward topic. Juju automatically retries after version 2.0 but you still have to manually retry in earlier versions. If someone has better words I am interested in the assistance.

Also it should be noted the `authors-*` docs are older and replaced with the `developer-*` docs, but this fix includes changes to both.

### CODE REVIEW

**Comment by mbruzek** [comment]
@cmars please take a look and tell me if you have better words.

File: src/en/authors-hook-debug.md [hunk 1]
```diff
@@ -92,8 +92,10 @@ after the unit comes up, making it difficult to start a debug-hooks session in
 time to intercept them. If you're having difficulties, you can temporarily
 return an error code from your `install` hook (e.g. add an `exit 1` at the end
 of it), and start your session only when the unit reports an [error status
-](./authors-hook-errors.html). You should then run `juju resolved --retry` for
-the affected unit, and go back to the debug-hooks session to interact.
+](./authors-hook-errors.html). Juju version 2.0 and up will automatically retry
```
Comment (cmars) [inline src/en/authors-hook-debug.md]: When I develop charms with Juju 2.0, I usually don't wait around for the hook to retry though. How about something like:

> Hooks in error must be retried manually in Juju releases prior to 2.0, by issuing the command `juju resolved --retry unit-name\#` for the affected unit. Juju version 2.0 and up will automatically retry hooks in error periodically. However, `juju resolved --retry` may still be used to retry the hook immediately. After retrying, go back to the debug-hooks session to interact with the Juju environment.

I'm probably not fit to review `juju debug-hooks` in particular. I never use it; I just use print statements to debug things and use `charm build ...; juju upgrade-charm ...; juju resolved --retry ...` to try changes out iteratively.

File: src/en/authors-hook-debug.md [hunk 2]
```diff
@@ -71,9 +71,9 @@ queue until you exit your current window. See the  special considerations below.
 The queue for pending hooks will restart once you exit the window with an `exit`
 command.
 
-**Note:** To allow Juju to continue processing events normally, you **must**
-exit the hook execution window with the `exit` command, otherwise all further
-events on that unit will be paused indefinitely.
+!!! Note: To allow Juju to continue processing events normally, you **must**
+exit the hook execution with a zero return code (using the `exit` command),
+otherwise all further events on that unit may be paused indefinitely.
```
Comment (cmars) [inline src/en/authors-hook-debug.md]: s/paused/blocked/ ?

**Comment by cmars** [comment]
Some suggestions.

File: src/en/authors-hook-debug.md (hunk 2 above)
Comment (evilnick) [inline src/en/authors-hook-debug.md]: yeah, I think blocked is better

File: src/en/authors-hook-debug.md (hunk 1 above)
Comment (evilnick) [inline src/en/authors-hook-debug.md]: I recall there was discussion around some sort of upper limit to automatic retries. Did that ever happen?

File: src/en/developer-debugging.md [hunk 3]
```diff
@@ -74,14 +74,18 @@ check the `db-relation-joined` and `db-relation-broken` hooks:
 juju debug-hooks mysql/0 db-relation-joined db-relation-broken
 ```
 
-Additionally, if your hook is already in an error state, you can tell Juju to
-retry the hook after running `juju debug-hooks`:
+Additionally, if your hook is already in an error state, you can enter in an
```
Comment (evilnick) [inline src/en/developer-debugging.md]: enter into

File: src/en/troubleshooting-upgrade.md [hunk 4]
```diff
@@ -53,5 +53,8 @@ sometimes lead to hook failures. Connect to that unit, see what is wrong, and
 retry with:
 
 ```bash
-juju resolved --retry
+juju resolved --retry unit-name/#
 ```
+
+Where unit-name is replaced with the name of the unit, and # is replaced with
+the unit number you wish to restart.
```
Comment (evilnick) [inline src/en/troubleshooting-upgrade.md]: could possible do with an example, e.g.

   juju resolved --retry mariadb/1

File: src/en/authors-hook-debug.md (hunk 1 above)
Comment (cmars) [inline src/en/authors-hook-debug.md]: Not sure. I'll ask the mailing list.

Comment (cmars) [inline src/en/authors-hook-debug.md]: @evilnick To answer your question.. there is no upper limit, but there is a backoff delay. See https://lists.ubuntu.com/archives/juju-dev/2016-June/005716.html

Comment (cmars) [inline src/en/authors-hook-debug.md]: Also note that automatic retry can be disabled: https://lists.ubuntu.com/archives/juju-dev/2016-June/005718.html

Comment (evilnick) [inline src/en/authors-hook-debug.md]: thanks for that, I will add another issue to make sure this is captured in other relevant places

File: src/en/authors-hook-debug.md (hunk 2 above)
Comment (mbruzek) [inline src/en/authors-hook-debug.md]: I am happy to fix that, but want to point out that is the only part of the section that I did not change.

**Comment by evilnick** [comment]
@mbruzek I'm not sure what you did in your last commit but there now seem to be conflicts :(

**Comment by evilnick** [comment]
:fireworks: :tada: :balloon: :candy: :icecream:

**Pull request merged**